cache/save_skip.go

```diff
@@ -1,48 +1,100 @@
 package cache
 
 import (
-	"fmt"
 	"strings"
 )
 
-func (s *saver) canSkipSave(keyTemplate, evaluatedKey string, onlyCheckCacheKey bool) (canSkip bool, reason string) {
+type skipReason int
+
+const (
+	reasonKeyNotDynamic skipReason = iota
+	reasonNoRestore
+	reasonRestoreSameUniqueKey
+	reasonRestoreSameKeyNotUnique
+	reasonRestoreOtherKey
+	reasonNewArchiveChecksumMatch
+	reasonNewArchiveChecksumMismatch
+)
+
+func (r skipReason) String() string {
+	switch r {
+	case reasonKeyNotDynamic:
+		return "key_not_dynamic"
+	case reasonNoRestore:
+		return "no_restore_found"
+	case reasonRestoreSameUniqueKey:
+		return "restore_same_unique_key"
+	case reasonRestoreSameKeyNotUnique:
+		return "restore_same_key_not_unique"
+	case reasonRestoreOtherKey:
+		return "restore_other_key"
+	case reasonNewArchiveChecksumMatch:
+		return "new_archive_checksum_match"
+	case reasonNewArchiveChecksumMismatch:
+		return "new_archive_checksum_mismatch"
+	default:
+		return "unknown"
+	}
+}
+
+func (r skipReason) description() string {
+	switch r {
+	case reasonKeyNotDynamic:
+		return "key is not dynamic; the expectation is that the same key is used for saving different cache contents over and over"
+	case reasonNoRestore:
+		return "no cache was restored in the workflow, creating a new cache entry"
+	case reasonRestoreSameUniqueKey:
+		return "a cache with the same key was restored in the workflow, new cache would have the same content"
+	case reasonRestoreSameKeyNotUnique:
+		return "a cache with the same key was restored in the workflow, but contents might have changed since then"
+	case reasonRestoreOtherKey:
+		return "there was no cache restore in the workflow with this key"
+	case reasonNewArchiveChecksumMatch:
+		return "new cache archive is the same as the restored one"
+	case reasonNewArchiveChecksumMismatch:
+		return "new cache archive doesn't match the restored one"
+	default:
+		return "unrecognized skipReason"
+	}
+}
+
+func (s *saver) canSkipSave(keyTemplate, evaluatedKey string, isKeyUnique bool) (bool, skipReason) {
 	if keyTemplate == evaluatedKey {
-		return false, "key is not dynamic; the expectation is that the same key is used for saving different cache contents over and over"
+		return false, reasonKeyNotDynamic
 	}
 
 	cacheHits := s.getCacheHits()
 	if len(cacheHits) == 0 {
-		return false, "no cache was restored in the workflow, creating a new cache entry"
+		return false, reasonNoRestore
 	}
 
 	if _, ok := cacheHits[evaluatedKey]; ok {
-		if onlyCheckCacheKey {
-			return true, "a cache with the same key was restored in the workflow, new cache would have the same content"
+		if isKeyUnique {
+			return true, reasonRestoreSameUniqueKey
 		} else {
-			return false, "a cache with the same key was restored in the workflow, but contents might have changed since then"
+			return false, reasonRestoreSameKeyNotUnique
 		}
 	}
 
-	otherKeys := []string{}
```

Comment:
Moved this to a separate function and let `save.go` call this because we can't encode this extra info into the enum.